insights/combiners/krb5.py

```diff
@@ -0,0 +1,115 @@
+"""
+krb5 configuration
+==================
+The krb5 files are normally available to rules as a list of
+Krb5Configuration objects.
+"""
+
+from .. import LegacyItemAccess
+from insights.core.plugins import combiner
+from insights.parsers.krb5 import Krb5Configuration
+
+
+@combiner(requires=[Krb5Configuration])
+class AllKrb5Conf(LegacyItemAccess):
+    """
+    Combiner for accessing all the krb5 configuration files, the format is dict.
+    There may be multi files for krb5 configuration, and the main config file is
+    krb5.conf. In the situation that same section is both in krb5.conf and other
+    configuration files, section in krb5.conf is the available setting. Data from
+    parser krb5 is list of dict(s), this combiner will parse this list and return
+    a dict which containing all valid data.
+
+    Sample files::
+
+        /etc/krb5.conf:
+
+            includedir /etc/krb5.conf.d/
+            include /etc/krb5test.conf
+            module /etc/krb5test.conf:residual
+
+            [logging]
+                default = FILE:/var/log/krb5libs.log
+                kdc = FILE:/var/log/krb5kdc.log
+
+        /etc/krb5.d/krb5_more.conf:
+
+            [logging]
+                default = FILE:/var/log/krb5.log
+                kdc = FILE:/var/log/krb5.log
+                admin_server = FILE:/var/log/kadmind.log
+
+            [realms]
+                dns_lookup_realm = false
+                default_ccache_name = KEYRING:persistent:%{uid}
+
+    Examples:
+        >>> all_krb5 = shared[AllKrb5Conf]
+        >>> all_krb5.include
+        ['/etc/krb5test.conf']
+        >>> all_krb5.sections()
+        ['logging', 'realms']
+        >>> all_krb5.options('logging')
+        ['default', 'kdc']
+        >>> all_krb5['logging']['kdc']
+        'FILE:/var/log/krb5kdc.log'
+        >>> all_krb5.has_option('logging', 'admin_server')
```

Comment:
Unfortunately this behaviour is not clearly defined.  `man krb5.conf` says:
```
The krb5.conf file can include other files using either of the following directives at the beginning of a line:

          include FILENAME
          includedir DIRNAME

FILENAME  or  DIRNAME should be an absolute path. The named file or directory must exist and be readable.  Including a directory includes all files within the directory whose names consist solely of alphanumeric characters, dashes, or underscores.  Included profile files are syntactically independent of their parents, so each included file must begin with a section header.
```
I take the point about 'included profile files [being] syntactically independent of their parents' to mean that the files are read independently and then merged, rather than being read as if the included file was the literal content of the original file (as the `#include` directive in C does, for example).  However, I see no evidence there or in discussion about this online to decide whether the `/etc/krb5.conf` file takes precedence or not, and whether files are read in name order or directory order.

Just as a note, it's worth noting that the `include` and `includedir` directives can occur outside a section heading.

One way of resolving this is simply to test what happens in Kerberos.  Does it allow multiple files to redefine the same section?  The same key?  What happens?  Have we tried this?  Can we?

If we can't verify how Kerberos works in the short term, we may have to make a decision and modify it later if it turns out to be incorrect.  OTOH, I'd suggest that we adopt the reverse approach to that taken in the code - I think we should read the main file first, and then override it with any included files.  This is slightly simpler, but it's also in keeping with the general process of include files overriding default behaviour.